daemon/system-info.sh

```diff
@@ -393,6 +393,46 @@ elif pgrep "kubelet"; then
   HOST_IS_K8S_NODE="true"
 fi
 
+# ------------------------------------------------------------------------------------------------
+# Detect instance metadata for VMs running on cloud providers
+
+CLOUD_TYPE="unknown"
+CLOUD_INSTANCE_TYPE="unknown"
+CLOUD_INSTANCE_REGION="unknown"
+
+if [ "${VIRTUALIZATION}" != "none" ] && command -v curl > /dev/null 2>&1; then
+  # Try AWS IMDSv2
+  if [ "${CLOUD_TYPE}" = "unknown" ]; then
+    AWS_IMDS_TOKEN="$(curl -s -m 5 --noproxy -X PUT "http://169.254.169.254/latest/api/token" -H "X-aws-ec2-metadata-token-ttl-seconds: 21600")"
+    if [ -n "${AWS_IMDS_TOKEN}" ]; then
+      CLOUD_TYPE="AWS"
+      CLOUD_INSTANCE_TYPE="$(curl -s -m 5 --noproxy -H "X-aws-ec2-metadata-token: $TOKEN" -v "http://169.254.169.254/latest/meta-data/instance-type")"
+      CLOUD_INSTANCE_REGION="$(curl -s -m 5 --noproxy -H "X-aws-ec2-metadata-token: $TOKEN" -v "http://169.254.169.254/latest/meta-data/placement/region")"
+    fi
+  fi
+
+  # Try GCE computeMetadata v1
+  if [ "${CLOUD_TYPE}" = "unknown" ]; then
+    if [ -n "$(curl -s -m 5 --noproxy -H "Metadata-Flavor: Google" "http://metadata.google.internal/computeMetadata/v1")" ]; then
+      CLOUD_TYPE="GCP"
+      CLOUD_INSTANCE_TYPE="$(curl -s -m 5 --noproxy -H "Metadata-Flavor: Google" "http://metadata.google.internal/computeMetadata/v1/instance/machine-type")"
```

Comment:
```suggestion
    if [ -n "$(curl --fail -s -m 5 --noproxy "*" -H "Metadata-Flavor: Google" "http://metadata.google.internal/computeMetadata/v1")" ]; then
      CLOUD_TYPE="GCP"
      CLOUD_INSTANCE_TYPE="$(curl --fail -s -m 5 --noproxy "*" -H "Metadata-Flavor: Google" "http://metadata.google.internal/computeMetadata/v1/instance/machine-type")"
      [ -n "$CLOUD_INSTANCE_TYPE" ] && CLOUD_INSTANCE_TYPE=$(basename "$CLOUD_INSTANCE_TYPE")
      CLOUD_INSTANCE_REGION="$(curl --fail -s -m 5 --noproxy "*" -H "Metadata-Flavor: Google" "http://metadata.google.internal/computeMetadata/v1/instance/zone")"
      [ -n "$CLOUD_INSTANCE_REGION" ] && CLOUD_INSTANCE_REGION=$(basename "$CLOUD_INSTANCE_REGION")
```

- Problem 1:
  - description: `--noproxy` expects a parameter, it doesn't work correctly without it (I guess it uses `-H` as a parameter).
  - solution: use `--noproxy "*"`
  - a note: why do we need this parameter at all?

<details><summary>see the output with the current implementation </summary>
```cmd
bash-5.1# curl -s -m 5 --noproxy -H "Metadata-Flavor: Google" "qq" "http://metadata.google.internal/computeMetadata/v1/instance/zone"
<!DOCTYPE html>
<html lang=en>
  <meta charset=utf-8>
  <meta name=viewport content="initial-scale=1, minimum-scale=1, width=device-width">
  <title>Error 403 (Forbidden)!!1</title>
  <style>
    *{margin:0;padding:0}html,code{font:15px/22px arial,sans-serif}html{background:#fff;color:#222;padding:15px}body{margin:7% auto 0;max-width:390px;min-height:180px;padding:30px 0 15px}* > body{background:url(//www.google.com/images/errors/robot.png) 100% 5px no-repeat;padding-right:205px}p{margin:11px 0 22px;overflow:hidden}ins{color:#777;text-decoration:none}a img{border:0}@media screen and (max-width:772px){body{background:none;margin-top:0;max-width:none;padding-right:0}}#logo{background:url(//www.google.com/images/branding/googlelogo/1x/googlelogo_color_150x54dp.png) no-repeat;margin-left:-5px}@media only screen and (min-resolution:192dpi){#logo{background:url(//www.google.com/images/branding/googlelogo/2x/googlelogo_color_150x54dp.png) no-repeat 0% 0%/100% 100%;-moz-border-image:url(//www.google.com/images/branding/googlelogo/2x/googlelogo_color_150x54dp.png) 0}}@media only screen and (-webkit-min-device-pixel-ratio:2){#logo{background:url(//www.google.com/images/branding/googlelogo/2x/googlelogo_color_150x54dp.png) no-repeat;-webkit-background-size:100% 100%}}#logo{display:inline-block;height:54px;width:150px}
  </style>
  <a href=//www.google.com/><span id=logo aria-label=Google></span></a>
  <p><b>403.</b> <ins>That’s an error.</ins>
  <p>Your client does not have permission to get URL <code>/computeMetadata/v1/instance/zone</code> from this server. Missing Metadata-Flavor:Google header. <ins>That’s all we know.</ins>
```
</details>

- Problem 2:
  - description: on unsucessfull response (not 200) we get output written to stdout (not stderr).
  - solution: add `--fail` parameter.

<details><summary>see the output w/ and w/o the option </summary>

- good

```bash
bash-5.1# curl --fail -s -m 5 --noproxy "*" -H "Metadata-Flavor: Google" "qq" "http://metadata.google.internal/computeMetadata/v1/instance/WRONG"
bash-5.1#
```

- bad

```bash
bash-5.1# curl -s -m 5 --noproxy "*" -H "Metadata-Flavor: Google" "qq" "http://metadata.google.internal/computeMetadata/v1/instance/WRONG"
<!DOCTYPE html>
<html lang=en>
  <meta charset=utf-8>
  <meta name=viewport content="initial-scale=1, minimum-scale=1, width=device-width">
  <title>Error 404 (Not Found)!!1</title>
  <style>
    *{margin:0;padding:0}html,code{font:15px/22px arial,sans-serif}html{background:#fff;color:#222;padding:15px}body{margin:7% auto 0;max-width:390px;min-height:180px;padding:30px 0 15px}* > body{background:url(//www.google.com/images/errors/robot.png) 100% 5px no-repeat;padding-right:205px}p{margin:11px 0 22px;overflow:hidden}ins{color:#777;text-decoration:none}a img{border:0}@media screen and (max-width:772px){body{background:none;margin-top:0;max-width:none;padding-right:0}}#logo{background:url(//www.google.com/images/branding/googlelogo/1x/googlelogo_color_150x54dp.png) no-repeat;margin-left:-5px}@media only screen and (min-resolution:192dpi){#logo{background:url(//www.google.com/images/branding/googlelogo/2x/googlelogo_color_150x54dp.png) no-repeat 0% 0%/100% 100%;-moz-border-image:url(//www.google.com/images/branding/googlelogo/2x/googlelogo_color_150x54dp.png) 0}}@media only screen and (-webkit-min-device-pixel-ratio:2){#logo{background:url(//www.google.com/images/branding/googlelogo/2x/googlelogo_color_150x54dp.png) no-repeat;-webkit-background-size:100% 100%}}#logo{display:inline-block;height:54px;width:150px}
  </style>
  <a href=//www.google.com/><span id=logo aria-label=Google></span></a>
  <p><b>404.</b> <ins>That’s an error.</ins>
  <p>The requested URL <code>/computeMetadata/v1/instance/WRONG</code> was not found on this server.  <ins>That’s all we know.</ins>
```

</details>

- Problem 3:
  - description: returned values are like `projects/902917867848/machineTypes/g1-small`.
  - solution: use `basepath`.

Reply:
I think we can use `http://metadata.google.internal/computeMetadata/v1/instance/zone` for `CLOUD_INSTANCE_REGION`. I couldn't find how to get the region, but we can extract it from the zone.

>  For example, the us-west1 region denotes a region on the west coast of the United States that has three zones: us-west1-a, us-west1-b, and us-west1-c.

I got "us-central1-a", so we can just remove the last part after `-`.

Reply:
I tested the code I'm suggesting on a GKE

```bash
bash-5.1# cat test.sh
if [ -n "$(curl --fail -s -m 5 --noproxy "*" -H "Metadata-Flavor: Google" "http://metadata.google.internal/computeMetadata/v1")" ]; then
  CLOUD_TYPE="GCP"
  CLOUD_INSTANCE_TYPE="$(curl --fail -s -m 5 --noproxy "*" -H "Metadata-Flavor: Google" "http://metadata.google.internal/computeMetadata/v1/instance/machine-type")"
  [ -n "$CLOUD_INSTANCE_TYPE" ] && CLOUD_INSTANCE_TYPE=$(basename "$CLOUD_INSTANCE_TYPE")
  CLOUD_INSTANCE_REGION="$(curl --fail -s -m 5 --noproxy "*" -H "Metadata-Flavor: Google" "http://metadata.google.internal/computeMetadata/v1/instance/zone")"
  [ -n "$CLOUD_INSTANCE_REGION" ] && CLOUD_INSTANCE_REGION=$(basename "$CLOUD_INSTANCE_REGION") && CLOUD_INSTANCE_REGION=${CLOUD_INSTANCE_REGION%-*}

  echo "$CLOUD_TYPE" "$CLOUD_INSTANCE_TYPE" "$CLOUD_INSTANCE_REGION"
fi

bash-5.1# ./test.sh
GCP g1-small us-central1
```

Reply:
1. We need it because this is a specially handled address in the link-local scope. Routing this through a proxy does not work correctly (just like trying to route a loopback address would not work correctly).
2. Agreed, I completely forgot about this behavior of cURL.
3. I had not thought of that approach.

And, regarding the region, you appear to be correct.